relay-server/src/actors/events.rs

```diff
@@ -1387,9 +1396,112 @@ impl Handler<QueueEnvelope> for EventManager {
     }
 }
 
+/// Takes an envelope and potentially removes the transaction item from it if that
+/// transaction item should be sampled out according to the dynamic sampling configuration
+/// and the trace context.
+fn sample_transaction_internal<'a>(
+    envelope: &'a mut Envelope,
+    project_state: Option<&ProjectState>,
+) -> &'a mut Envelope {
+    let project_state = match project_state {
+        None => return envelope,
+        Some(project_state) => project_state,
+    };
+
+    let sampling_config = match project_state.config.sampling_config {
+        // without sampling config we cannot sample transactions so give up here
+        None => return envelope,
+        Some(ref sampling_config) => sampling_config,
+    };
+
+    let project_id = match project_state.project_id {
+        None => return envelope,
+        Some(project_id) => project_id,
+    };
+
+    let trace_context = envelope.get_trace_context();
+    let transaction_item = envelope.get_item_by(|item| item.ty() == ItemType::Transaction);
+
+    let trace_context = match (trace_context, transaction_item) {
+        // we don't have what we need, can't sample the transactions in this envelope
+        (None, _) | (_, None) => return envelope,
+        // see if we need to sample the transaction
+        (Some(trace_context), Some(_)) => trace_context,
+    };
+
+    let should_sample = trace_context
+        // see if we should sample
+        .should_sample(sampling_config, project_id)
+        // TODO verify that this is the desired behaviour (i.e. if we can't find a rule
+        // for sampling, include the transaction)
+        .unwrap_or(true);
+
+    if !should_sample {
+        // finally we decided that we should sample the transaction
+        envelope.take_item_by(|item| item.ty() == ItemType::Transaction);
+    }
+
+    envelope
+}
+
+/// Check if we should remove transactions from this envelope (because of trace sampling) and
+/// return what is left of the envelope
+pub fn sample_transaction(
```

Comment:
This function seems a bit out-of-place in the events actor file. Can we move this to utils? 

I'm assuming it hasn't been moved there due to the dependency on the Project actor. We have other utilities with dependencies on actors, but if you don't like that you can also move it into the project actor's file.

Reply:
It was placed in events.rs because that is where it is used (mostly :-) ). I had to make it public because I also needed it in endpoints/common.rs ( and since common already imports from events.rs it didn't look to bad).

Moved it into `relay_server::utils::dynamic_sampling`.